### PR TITLE
expr: Use checked recursion in could_run_expensive_function

### LINKED ISSUE #20551: Use checked recursion or iteration for auto-routing simple queries to mz_introspection
Replace the MIR expression evaluation recursion with checked recursion or iteration, as added in https://github.com/MaterializeInc/materialize/pull/20372.

See https://materializeinc.slack.com/archives/C02FWJ94HME/p1689188402903309?thread_ts=1689149048.591139&cid=C02FWJ94HME for context.
Rationale

> Evaluating an MIR expression with a recursive function is potentially dangerous because depending on the MIR structure, it's easy to overflow the stack. It's better to use an iterative algorithm if possible, or guard the stack so it won't crash at runtime.

### PR DESCRIPTION
**Pull request by andrewrodriguez-m**
### Motivation

Follow up on review comments in #20372. Fixes #20551.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/expr/src/relation/mod.rs [hunk 1]
```diff
@@ -1789,33 +1789,28 @@ impl MirRelationExpr {
...
-            result = result || scalar_could_run_expensive_function(scalar)
-        });
-        self.visit_children(|e| result = result || e.could_run_expensive_function());
-        result
+        self.try_could_run_expensive_function().unwrap_or(true)
+    }
+    fn try_could_run_expensive_function(&self) -> Result<bool, RecursionLimitError> {
+        let mut result = false;
+        self.try_visit_pre(&mut |e| {
+            // FlapMap has a table function while all other constructs use MirScalarExpr to run a function
```
[the remark this answers is not in the document]
Comment (aalexandrov) [reply src/expr/src/relation/mod.rs:1797]: > while all other constructs use MirScalarExpr to run a function

Strictly speaking this is not true, as `Reduce` uses `AggregateExpr` which runs `AggregateFunc` to define aggregations.

[the remark this answers is not in the document]
Comment (andrewrodriguez-m) [reply src/expr/src/relation/mod.rs]: thanks. fixed both

**Review by aalexandrov — APPROVED** [review]
Some stylistic and API suggestions, otherwise LGTM.

File: src/expr/src/relation/mod.rs [hunk 2]
```diff
@@ -1789,33 +1789,33 @@ impl MirRelationExpr {
     /// The goal is to be a heuristic for "this expression is cheap to evaluate:
```
[the remark this answers is not in the document]
Comment (andrewrodriguez-m) [reply src/expr/src/relation/mod.rs]: done. replaced my comment with yours

File: src/expr/src/relation/mod.rs [hunk 3]
```diff
@@ -1789,33 +1789,33 @@ impl MirRelationExpr {
...
-        let mut result = match self {
-            MirRelationExpr::FlatMap { .. } => true,
-            _ => false,
-        };
-        self.visit_scalars(&mut |scalar| {
-            result = result || scalar_could_run_expensive_function(scalar)
-        });
-        self.visit_children(|e| result = result || e.could_run_expensive_function());
-        result
+        self.try_could_run_expensive_function().unwrap_or(true)
```
[the remark this answers is not in the document]
Comment (andrewrodriguez-m) [reply src/expr/src/relation/mod.rs]: looks good. took that code

**Pull request merged**